File: TiaCodegen/Commands/Functions/TONCall.cs
using System.Linq;
using TiaCodegen.Commands.Functions.Base;
using TiaCodegen.Enums;
using TiaCodegen.Interfaces;

namespace TiaCodegen.Commands.Functions
{
    public class TONCall : SystemFunctionBlockCall
    {
        public TONCall(string instanceName,
            IOperationOrSignal pt = null,
            IOperationOrSignal q = null,
            IOperationOrSignal et = null) : base("TON", instanceName, null)
        {
            Interface["PT"] = new IOperationOrSignalDirectionWrapper(pt, Direction.Input);
            Interface["Q"] = new IOperationOrSignalDirectionWrapper(q, Direction.Output);
            Interface["ET"] = new IOperationOrSignalDirectionWrapper(et, Direction.Output);

            TemplateValueName = "time_type";
            TemplateValueType = "Type";
            TemplateValue = "Time";
            HasNoEn = true;

            SafetyTemplateString = @"      <TemplateValue Name=""f_user_card"" Type=""Cardinality"">1</TemplateValue>
      <TemplateValue Name=""f_image_card"" Type=""Cardinality"">0</TemplateValue>
      <TemplateValue Name=""f_imageclassic_card"" Type=""Cardinality"">0</TemplateValue>
      <TemplateValue Name=""f_imageplus_card"" Type=""Cardinality"">0</TemplateValue>";

            Children.AddRange(Interface.Values.Where(x => x.OperationOrSignal != null).Select(x => x.OperationOrSignal));
        }
    }
}

File: TiaCodegen-ts/src/Commands/Functions/TONCall.ts
import { IOperationOrSignal } from '../../Interfaces/IOperationOrSignal.js';
import { IOperationOrSignalDirectionWrapper } from '../../Interfaces/IOperationOrSignalDirectionWrapper.js';
import { Direction } from '../../Enums/Direction.js';
import { SystemFunctionBlockCall } from './Base/SystemFunctionBlockCall.js';

export interface TONCallOptions {
    instanceName: string;
    pt?: IOperationOrSignal | null;
    q?: IOperationOrSignal | null;
    et?: IOperationOrSignal | null;
}

export class TONCall extends SystemFunctionBlockCall {
    constructor(options: TONCallOptions) {
        const { instanceName, pt = null, q = null, et = null } = options;
        super({ functionName: 'TON', instanceName });
        this.iface['PT'] = new IOperationOrSignalDirectionWrapper(pt, Direction.Input);

        this.iface['Q'] = new IOperationOrSignalDirectionWrapper(q, Direction.Output);
        this.iface['ET'] = new IOperationOrSignalDirectionWrapper(et, Direction.Output);

        this.templateValueName = 'time_type';
        this.templateValueType = 'Type';
        this.templateValue = 'Time';
        this.hasNoEn = true;

        this.safetyTemplateString = `      <TemplateValue Name="f_user_card" Type="Cardinality">1</TemplateValue>
      <TemplateValue Name="f_image_card" Type="Cardinality">0</TemplateValue>
      <TemplateValue Name="f_imageclassic_card" Type="Cardinality">0</TemplateValue>
      <TemplateValue Name="f_imageplus_card" Type="Cardinality">0</TemplateValue>
      <TemplateValue Name="codedbool_type" Type="Type">DInt</TemplateValue>`;

        for (const w of Object.values(this.iface)) {
            if (w.operationOrSignal !== null) this.children.push(w.operationOrSignal);
        }
    }
}
